teps/0143-concise-parameters-results.md

```diff
@@ -0,0 +1,848 @@
+
+```yaml
+# before
+results:
+  - name: hello
+    type: string
+    value: world
+  - name: environments
+    type: array
+    value:
+      - 'staging'
+      - 'qa'
+      - 'prod'
+  - name: git
+    type: object
+    value:
+      url: abc.com
+      commit: sha123
+
+# after
+output:
+  results:
+    hello: world
+    environments:
+      - 'staging'
+      - 'qa'
+      - 'prod'
+    git:
+      url: abc.com
+      commit: sha123
+```
+
+### API
+
+#### Parameters
+
+```go
+// Add the new params
+type ParamSpecsMap map[string]ParamSpec
+type ParamsMap map[string]ParamValue
+
+// Add Inputs.Params and Outputs.Results
+type InputSpecs struct{
+  ParamSpecs ParamSpecsMap `json:"params,omitempty"`
+}
+type Inputs struct{
+  Params ParamsMap `json:"params,omitempty"`
+}
+```
+
+#### Results
+
+```go
+// Add the new results
+type ResultsSpecsMap map[string]ResultSpec
+type ResultsMap map[string]ResultValue
+
+// Use the ResultSpec to align current TaskResult and PipelineResult
+type ResultSpec struct{
+  Name string  `json:"name"`
+  Type ResultsType `json:"type,omitempty"`
+  Description string `json:"description,omitempty"`
+  Properties map[string]PropertySpec `json:"properties,omitempty"`
+  Value ResultValue  `json:"value"`
+}
+
+type OutputSpecs struct{
+  ResultSpecs ResultsSpecsMap `json:"results,omitempty"`
+}
+type Outputs struct{
+  ResultsMap map[string]ResultValue `json:"results,omitempty"`
+}
+```
+
+#### Task and TaskRun
+
+```go
+// Add Input and Output to tasks, pipelines, etc.
+// authoring time
+type TaskSpec struct{
+  Inputs InputSpecs `json:"inputs,omitempty"`
+  Outputs OutputSpecs `json:"outputs,omitempty"`
+}
+// runtime
+type TaskRunSpec struct{
+  Inputs Inputs `json:"inputs,omitempty"`
+}
+type TaskRunStatusFields struct{
+  Outputs Outputs `json:"outputs,omitempty"`
+}
+```
+
+#### Pipeline and PipelineRun
+
+```go
+// authoring time
+type PipelineSpec struct{
+  Inputs InputSpecs `json:"inputs,omitempty"`
+  Outputs OutputSpecs `json:"outputs,omitempty"`
+}
+// runtime
+type PipelineRunSpec struct{
+  Inputs Inputs `json:"inputs,omitempty"`
+}
+type PipelineRunStatusFields struct{
+  Outputs Outputs `json:"outputs,omitempty"`
+}
+```
+
+### Validation
+
+The new fields will be gated by a new feature flag, `enable-concise-api`, which defaults to `"false"` while the feature is in alpha.
+
+The validation webhook will validate either the map syntax or the current list syntax is used, users cannot use both at the same time.
+
+### Resolvers and Matrix
+
+`Resolvers` and `Matrix` use `Params` which can be updated to use the concise syntax. Since `Matrix` is an alpha feature, it can be updated with notice. `Resolvers` is beta feature so we need to promote the feature to beta before adopting the new syntax with a deprecation notice.
```

Comment:
> It can be updated with notice 
Did you mean without? Or with no need for deprecation?

> `Resolvers` is beta feature so we need to promote the feature to beta before adopting the new syntax with a deprecation notice.

I'm not sure I understand what you mean. Which feature has to be promoted to beta first?

Reply:
I was thinking about adding the new syntax in v1 and how we can update the current places where we use params (resolvers, matrix). If we choose to add the syntax in a2alpha1 then we don't need this section. I will update the proposal soon